### ISSUE TITLE
Karma opa5 plugin

### ISSUE THREAD

**Issue by geekflyer** [SAP/karma-openui5#1]
Hi @matz3 @TobiasOetzel,

would it be possible to develop an karma opa5 plugin (maybe based on the [qunit plugin](https://github.com/karma-runner/karma-qunit))? 
I want to run the opa5 tests in an automated fashion and get back structured results via tha karma junit reporter or the CLI. With a karma opa5 plugin we could easily benefit from the existing reporters etc.

Best Regards Christian

**Comment by cevou** [comment]
Hi Christian,

it is already possible to run OPA5 tests with karma and the karma OpenUI5 plugin. Here is an example configuration:

```javascipt
module.exports = function(config) {
	config.set({
		basePath: '',
		frameworks: ['qunit','openui5'],
		files: [
			{ pattern: 'bower_components/openui5-sap.ui.core/resources/**/*', served: true, included: false, watched: false },
			{ pattern: 'bower_components/openui5-sap.m/resources/**/*', served: true, included: false, watched: false },
			{ pattern: 'bower_components/openui5-themelib_sap_bluecrystal/resources/**/*', served: true, included: false, watched: false },
			{ pattern: 'src/**/*', served: true, included: false, watched: true },
			{ pattern: 'test/**/*.qunit.js', served: true, included: true, watched: true }
		],
		preprocessors: {
			'src/**/*.js': ['coverage']
		},
		openui5: {
			path: 'bower_components/openui5-sap.ui.core/resources/sap-ui-core.js',
			useMockServer: false
		},
		client: {
			openui5: {
				config: {
					theme: 'sap_bluecrystal',
					libs: 'sap.m',
					resourceroots: {
						'sap.m': '/base/bower_components/openui5-sap.m/resources/sap/m',
						'my.namespace': '/base/src/my/namespace'
					},
					themeroots: {
						'sap_bluecrystal': '/base/bower_components/openui5-themelib_sap_bluecrystal/resources'
					}
				}
			}
		},
		reporters: ['progress', 'coverage'],
		coverageReporter: {
			type : 'html',
			dir : 'coverage'
		},
		port: 9876,
		colors: true,
		logLevel: config.LOG_INFO,
		autoWatch: true,
		browsers: ['Chrome'],
		singleRun: false
	});
};
```

What is done here?
* Include OpenUI5 and QUnit for the tests as frameworks
* Serve all OpenUI5 related files
* Serve all files forom the src dicrectory
* Include all test scripts

* Specify that code coverage should be enabled for all JS files in the source directory
* Configure the OpenUI5 plugin

To record the code coverage correctly, you cannot run the tests in an IFrame. However, you can create an arrangement, that creates a component inside the karma test page:

```javascript
sap.ui.define([
	"jquery.sap.global",
	"sap/ui/test/Opa5",
	"sap/ui/core/ComponentContainer",
], function(jQuery, Opa5, ComponentContainer) {
	
	var oContainer = null;
	var oComponent = null;
	
	return Opa5.extend("my.namespace.test.opa.Arrangements", {
		iStartAComponent: function (sName) {
			if (!oContainer) {
				$body = jQuery("body");
				$body.addClass("sapUiBody").attr("role", "application");
				if ($body.find("#content").length === 0) {
					$body.append('<div id="content"></div>');
				}
				oContainer = new ComponentContainer();
				oContainer.placeAt("content");
			}
			if (oComponent) {
				oComponent.destroy();
				Opa5.getPlugin().mViews = {};
				oComponent = null;
			}
			oComponent = sap.ui.getCore().createComponent({
				name: sName,
			});
			oContainer.setComponent(oComponent);
			return this;
		}
	});	
});
```

Your test could look like this:

```javascript
sap.ui.define([
	"jquery.sap.global",
	"sap/ui/test/Opa5",
	"sap/ui/test/opaQunit",
	"my/namespace/test/opa/Arrangements"
], function(jQuery, Opa5, opaQunit, Arrangements) {

	Opa5.extendConfig({
		arrangements : new Arrangements(),
		viewNamespace : "my.namespace.app.view."
	});
	
	opaTest("TestApp", function(Given, When, Then) {
		Given.iStartAComponent("my.namespace.app");
	});
});
```

I hope this helps!

Best regards,
Christoph

**Comment by TobiasOetzel** [comment]
Thanks a lot christoph!

**Comment by geekflyer** [comment]
Thanks a lot @cevou ! That's exactly what I was looking for. Perhaps you could add your instructions to a sub section of the README.md of this plugin, so that future visitors know about it too :)

Best Regards Christian

**Comment by jasper07** [comment]
@cevou I have an app with Karma running my Opa tests, my app is running via a grunt server on another port and i use the proxy config so the test run.

I want to move over to using your plugin, less moving parts. A couple of changes to your code above and it works. However i cannot get the mockserver logic working, when the component sets up the model it still tries to retrieve the live service metadata.

my app uses the parameter ResponderOn=true logic to start the mockserver

is your mockeserver logic meant to be a replacement for ResponderOn? did you have issues setting the needed parameter? 

cannot see an issue relating to //workaround for crossroads/mockserver issue

also how would you add you own mockrequests?

**Comment by cevou** [comment]
Basically the `useMockServer` does the same as the `responderOn` logic in the tdg App. I haven't had any problems yet using the mockserver integration. Have you set `useMockServer` to `true` in the karma.conf.js?

**Comment by jasper07** [comment]
had time to debug and i got it working - had to make a couple of changes to mockserver.js to get it working

1 . the ui5config needed to go down another level
2.  the global var 'module' is undefined

```javascript
(function(window) {

var config = window.__karma__.config,
	ui5config = (config && config.openui5.config) || {},  //1
	mockserverConfig = ui5config.mockserver || {};

//workaround for crossroads/mockserver issue
(function() {
	if (typeof module === undefined){    //2
		var module = '';
	}
	var original = module;
	module = undefined;
	jQuery.sap.require("sap.ui.core.util.MockServer");
	module = original;
})();

mockserverConfig.config = mockserverConfig.config ||{};

sap.ui.core.util.MockServer.config(mockserverConfig.config);

var oMockServer = new sap.ui.core.util.MockServer({
	rootUri: mockserverConfig.rootUri
});
oMockServer.simulate(mockserverConfig.metadataURL, mockserverConfig.mockdataSettings);
oMockServer.start();

})(window);
```

I copied the TDG app, to get it working i had to separate the app from the tests

I couldn't get the navigation tests working but that doesn't seem like a real scenario anywhere

```javascript
module.exports = function(config) {
    config.set({
        basePath: '',
        frameworks: ['qunit', 'openui5'],
        files: [{
            pattern: 'bower_components/openui5-sap.ui.core/resources/**/*',
            served: true,
            included: false,
            watched: false
        }, {
            pattern: 'bower_components/openui5-sap.m/resources/**/*',
            served: true,
            included: false,
            watched: false
        }, {
            pattern: 'bower_components/openui5-sap.ui.layout/resources/**/*',
            served: true,
            included: false,
            watched: false
        }, {
            pattern: 'bower_components/openui5-themelib_sap_bluecrystal/resources/**/*',
            served: true,
            included: false,
            watched: false
        }, {
            pattern: 'webapp/**/*',
            served: true,
            included: false,
            watched: true
        }, {
            pattern: 'test/**/*',
            served: true,
            included: true,
            watched: true
        }],
        // preprocessors: {
        //     'src/**/*.js': ['coverage']
        // },
        openui5: {
            path: 'bower_components/openui5-sap.ui.core/resources/sap-ui-core.js',
            useMockServer: true
        },
        client: {
            openui5: {
                config: {
                    theme: 'sap_bluecrystal',
                    libs: 'sap.m',
                    resourceroots: {
                        'sap.m': '/base/bower_components/openui5-sap.m/resources/sap/m',
                        'sap.ui.layout': '/base/bower_components/openui5-sap.ui.layout/resources/sap/ui/layout',
                        'sap.ui.demo.tdg': '/base/webapp',
                        'sap.ui.demo.tdg.test': '/base/test'
                    },
                    themeroots: {
                        'sap_bluecrystal': '/base/bower_components/openui5-themelib_sap_bluecrystal/resources'
                    },
                    mockserver: {
                        config: {
                            autoRespond: true
                        },
                        rootUri: '/uilib-sample/proxy/http/services.odata.org/V2/(S(sapuidemobespractices))/OData/OData.svc/',
                        metadataURL: '/base/webapp/model/metadata.xml',
                        mockdataSettings: '/base/webapp/model/'
                    }
                }
            }
        },
        reporters: ['progress'], //'coverage'],
        // coverageReporter: {
        //     type: 'html',
        //     dir: 'coverage'
        // },
        port: 9876,
        colors: true,
        logLevel: config.LOG_DEBUG,
        autoWatch: true,
        browsers: ['Chrome','PhantomJS'],
        singleRun: false
    });
};
```

**Comment by jasper07** [comment]
for those interested in a working example
https://github.com/jasper07/karmatest

**Comment by matz3** [comment]
I guess this got solved so I'm closing the issue

**Comment by 739823891** [comment]
Hi @janmattfeld 
I copied your code to my computer to test it,but I found it's unstable.I use the jenkins server to run it every two minute.In most cases,it runs successfully,but it may failed sometimes without anything changed,Jenkins reports:
 [1A[2KChrome 45.0.2454 (Windows 8.1 0.0.0): Executed 0 of 0 SUCCESS (0 secs / 0 secs)
[1A[2KChrome 45.0.2454 (Windows 8.1 0.0.0): Executed 0 of 0 ERROR (0.001 secs / 0 secs).
I run it 72 times,it return failure 10 times.
Do you have any idea to solve it?Thank you.

**Comment by matz3** [comment]
@739823891 could you please open a new issue with your problem?
As you log does not show any hint, please re-run it with `--log-level debug` (see [here](http://karma-runner.github.io/0.13/config/configuration-file.html) to get more output.

**Comment by amit1agrawal** [comment]
Hi @jasper07 ,
I copied your code in my computer to test it. I have done the changes as suggested by @matz3 that is to load 'openui5' framework before 'qunit' 
*frameworks: ['openui5','qunit']*
So there is no issue with loading of MockServer.js but mock data is not getting loaded. I have set the log level to DEBUG and nowhere I could see that the mock data json files are being loaded/served. Also, in chrome network log, there is an entry to load metadata.xml from actual service (which fails of course). 
Am I doing something wrong? I can provide the debug log files if required.

Thanks in advance.
